# TRESPI Search Demonstration

### System Setup

In [1]:
import trespi_demo as td

index_path = '/home/ubuntu/efs/msmarco_leaderboard_attempt8/lucene_idx_07232021'
msmarco_docs_path = '/home/ubuntu/efs/data/msmarco_docs/msmarco-docs.tsv'
trespi = td.TrespiDemo(index_path, msmarco_docs_path, k1=5.5, b=1)

Reading index from  /home/ubuntu/efs/data/msmarco_docs/msmarco-docs.tsv.pickled-index


# Query #1 -- Reaheating Lasagna
### MSMARCO Query from training set

According to MSMARCO, the relevant document for this query is D2286511. Let's run the query and view the results.

In [2]:
qresults = trespi.run_query('what temperature should you reheat lasagna to')
trespi.print_results(qresults, 'D2286511', 8)

| Pos | Reciprocol Rank  | Doc.ID  | Index Rank | Title |
| ----| ---------------- | ----- | ------ | ----- |
| 1 | 1.000 | D1262534 | 17.50 | What temperature would you heat frozen lasagna in the oven a
| 2 | 0.500 | D2340776 | 17.25 | how long should I reheat frozen lasagna?
| 3 | 0.333 | D1998010 | 17.01 | Reheating unfrozen lasagna (whole)
| 4 | 0.250 | D2286510 | 16.99 | Best Way to Reheat Lasagna
| 5 | 0.200 | D1521839 | 16.30 | How Do You Reheat Pizza in the Oven?
| 6 | 0.167 | **D2286511** | 16.25 | How long do you REHEAT LASAGNA for in the oven, when you mak
| 7 | 0.143 | D985758 | 16.19 | What Is the Best Way to Reheat Lasagna?
| 8 | 0.125 | D985755 | 15.89 | How Do I Reheat a Pan of Lasagna in the Oven?

The top eight search results are displayed. The document deemed most relevant by MSMARCO shows up in position #6.

Note the Reciprocol Rank column. It is the reciprocol of the document's position within the search results. The top document has reciprocol rank of 1, which is the maximum possible value. This query result is assigned a score of 0.167, which is the reciprocol rank of the MSMARCO relevant document.

### Review Relevant Document

In [3]:
trespi.print_qry_result(qresults, 6, text_slice=(900, 1400))

Query:		 what temperature should you reheat lasagna to
Document ID:	 D2286511
Position:	 6
Title:		 How long do you REHEAT LASAGNA for in the oven, when you make them the day before?
Contents:
lps get a really crispy top) before cooking in the middle of an oven for 50-60 minutes gas mark 4/180C/350F or until golden and bubbling. Loads of red wine and garlic bread (it's the future).spuertos · 1 decade ago0 0 Comment Asker's rating Reheat baked pasta dishes such as lasagna in a 350 degree oven, covered. Depending on how thick the dish is, it could take from 30 to 45 minutes to heat completely. Mylan H · 1 decade ago2 0 Comment Because it has eggs, you have to be careful refrigerating it


### What About the Top Document?
The top ranked document is also relevant to the query.

In [4]:
trespi.print_qry_result(qresults, 1, text_slice=(1000, 1500))

Query:		 what temperature should you reheat lasagna to
Document ID:	 D1262534
Position:	 1
Title:		 What temperature would you heat frozen lasagna in the oven at?
Contents:
 '11 at 0:40add a comment1 Answer active oldest votesup vote2down voteaccepted350F is the magic temp. Just about every casserole cooks at 350, and you cook it until it starts being bubbly which will probably between 30 minutes and an hour (in this case, you should check to make sure it's hot in the center as well, since you're starting from frozen). I'd cover it early, and then uncover it near the end so the top won't get dry.share improve this answeranswered May 31 '11 at 0:06Satanicpuppy12k 27


# Query #2 -- Contextual Spoken Language Understanding
### MSMARCO Query from training set

In [5]:
qresults2 = trespi.run_query('contextual spoken language understanding')
trespi.print_results(qresults2, 'D1350520', 8)

| Pos | Reciprocol Rank  | Doc.ID  | Index Rank | Title |
| ----| ---------------- | ----- | ------ | ----- |
| 1 | 1.000 | D1015337 | 12.60 | Contextualized language and decontextualized language
| 2 | 0.500 | D1219732 | 12.60 | Foreign Language Teaching Methods: Technology
| 3 | 0.333 | D2809829 | 12.51 | How To Understand Spoken French For Fluency
| 4 | 0.250 | D118379 | 12.31 | Contextualization
| 5 | 0.200 | **D1350520** | 12.23 | Context (language use)
| 6 | 0.167 | D1497397 | 12.05 | 7 Listening Skills for Understanding Spoken English
| 7 | 0.143 | D2809827 | 11.55 | Understanding Spoken French: A Practical Approach
| 8 | 0.125 | D665171 | 11.28 | Web Translation

MSMARCO Document is from Wikipedia

In [8]:
trespi.print_qry_result(qresults2, 5, (0, 1000))

Query:		 contextual spoken language understanding
Document ID:	 D1350520
Position:	 5
Title:		 Context (language use)
Contents:
"In semiotics, linguistics, sociology and anthropology, context refers to those objects or entities which surround a focal event, in these disciplines typically a communicative event, of some kind. Context is ""a frame that surrounds the event and provides resources for its appropriate interpretation"". [1]: 2–3 It is thus a relativistic concept, only definable with respect to some focal event, not independently. Contents [ hide ]1 In linguistics2 In linguistic anthropology3 Contextual variables4 See also5 References In linguistics [ edit]Verbal context refers to the text or speech surrounding an expression (word, sentence, or speech act ). Verbal context influences the way an expression is understood; hence the norm of not citing people out of context. Since much contemporary linguistics takes texts, discourses, or conversations as the object of analysis, the

Top TRESPI document focuses more on child development.

In [7]:
trespi.print_qry_result(qresults2, 1, (0, 1000))

Query:		 contextual spoken language understanding
Document ID:	 D1015337
Position:	 1
Title:		 Contextualized language and decontextualized language
Contents:
Contextualized language and decontextualized language Contextualized language and decontextualized language. What influence do the two language forms have on a child with language difficulty's learning, and why is it important to know the difference? The conversations children are expected to take part in - classroom discourse - are much different from the conversations they may have at home. Conversations at home are generally with familiar people in cosy surroundings in an environment they know well. In essence, the language is contextualized. That is, a child learns language skills in a familiar place, doing well known activities. An example would be a young child helping mum do the dishes. The child learns about plates, cups, saucers and other kitchen items by being immersed in and talking with his/her parent about their envi

# Mean Reciprocol Rank (MRR)
**Mean Reciprocol Rank:** The average of the reciprocol ranks for a set of queries.

**MRR@k:** Only the top *k* search results are reviewed. If the most relevant document does not appear in the top *k* search results, the reciprocol rank for that query is zero.
* **MRR@100**: Used by MSMARCO Leaderboard
* **MRR@10**: Common evaluation metric in research papers
